text/0098-store-multiple-replay-segments-in-a-single-blob.md

```diff
@@ -0,0 +1,282 @@
+     - https://cloud.google.com/kms/docs/envelope-encryption
+   - Related, contiguous byte ranges will be encrypted independently of the rest of the file.
+   - We will use KMS to manage our key-encryption-keys.
+   - Data-encryption-keys will be generated locally and will be unique.
+3. Parts will be tracked in a metadata table on an AlloyDB instance(s).
+   - A full table schema is provided in the **Proposal** section.
+   - AlloyDB was chosen because its a managed database with strong point-query performace.
+   - The metadata table will contain the key used to decrypt the byte range.
+4. On read, parts will be fetched without fetching the full file.
+   - More details are provided in the **Technical Details** section.
+
+# Proposal
+
+First, a new table called "file_part_byte_range" with the following structure is created:
+
+| id  | key | path     | start | stop  | dek      | kek_id | created_at          |
+| --- | --- | -------- | ----- | ----- | -------- | ------ | ------------------- |
+| 1   | A:0 | file.bin | 0     | 6241  | Aq3[...] | 1      | 2023-01-01T01:01:01 |
+| 2   | B:0 | file.bin | 6242  | 8213  | ppT[...] | 1      | 2023-01-01T01:01:01 |
+| 3   | A:1 | file.bin | 8214  | 12457 | 99M[...] | 1      | 2023-01-01T01:01:01 |
+
+- The key field is client generated identifier.
+  - It is not unique.
+  - The value of the key field should be easily computable by your service.
+  - In the case of Session Replay the key could be a concatenation of `replay_id` and `segment_id`.
+    - Illustrated above as `replay_id:segment_id`.
+    - Alternatively, a true composite key could be stored on a secondary table which contains a reference to the `id` of the `file_part_byte_range` row.
+- Path is the location of the blob in our bucket.
+- Start and stop are integers which represent the index positions in an inclusive range.
+  - This range is a contiguous sequence of related bytes.
+  - In other words, the entirety of the file part's encrypted data is contained within the range.
+- The "dek" column is the **D**ata **E**ncryption **K**ey.
+  - The DEK is the key that was used to encrypt the byte range.
+  - The key itself is encrypted by the KEK.
+    - **K**ey **E**ncryption **K**ey.
+  - Encryption is explored in more detail in the following sections.
+- The "kek_id" column contains the ID of the KEK used to encrypt the DEK.
+  - This KEK can be fetched from a remote **K**ey **M**anagement **S**ervice or a local database table.
+
+Notice each row in the example above points to the same file but with different start and stop locations. This implies that multiple, independent parts can be present in the same file. A single file can be shared by hundreds of different parts.
+
+Second, the Session Replay recording consumer will not commit blob data to Google Cloud Storage for each segment. Instead it will buffer many segments and flush them together as a single blob to GCS. Next it will make a bulk insertion into the database for tracking.
```

Comment:
How is this buffering process going to look like? 
It will have to rely on persistent state to prevent you from loosing chunks in case of failover before a file is fully ingested while parts have been already committed in Kafka.

Also a more high level concern with this. 
If the idea is to build a generic system reusable by other features, relying on a specific consumer to do the buffering has the important implication that Kafka is always going to be a moving part of your system. Have you considered doing the buffering separately.

Reply:
> How is this buffering process going to look like?

Like buffering would in our Snuba consumers.  You keep an array of messages/rows/bytes in memory.  When it comes time to flush you zip them together in some protocol specific format.  In this case the protocol is byte concatenation.

> It will have to rely on persistent state to prevent you from loosing chunks in case of failover before a file is fully ingested while parts have been already committed in Kafka.

If we assume that the replays consumer pushes the bytes to another, generic Kafka consumer which buffers the files before upload then the persistent state will be the log.  Upload failures can be re-run from the last committed offset.  Persistent failures would have to be handled as part of a DLQ and would require re-running.  Potentially introducing a significant amount of latency between (in the replays case) a billing outcome and the replay recording being made available.

Assuming this buffering/upload step exists inside our existing consumer (i.e. not a generic service) then offsets will not be committed until after the batch has been uploaded.

> Also a more high level concern with this.
If the idea is to build a generic system reusable by other features, relying on a specific consumer to do the buffering has the important implication that Kafka is always going to be a moving part of your system. Have you considered doing the buffering separately.

I have considered that.  The idea being process A buffers a bunch of file-parts/offsets before sending the completed object to permanent storage either through direct interaction, filestore, or Kafka intermediary (or any number of intermediaries).  The problem is then that each call site is responsible for the buffer which is the "hardest" part of the problem.

Kafka is an important part of this system in my mind.  When I wrote this document I relied on the guarantees it provides.  I think if this is a generic service made available to the company then publishing to the topic should be a simple process for integrators.  The fact that its Kafka internally is an irrelevant implementation detail (to the caller).